File: solutions/ocaml/137-cheapest-flights-within-k-stops.ml
(*
  Problem 137: Cheapest Flights Within K Stops (LeetCode 787)
  Difficulty: Med
  Language: OCaml
*)
let find_cheapest_price n flights src dst k =
  let inf = max_int in
  let prices = Array.init n (fun i -> if i = src then 0 else inf) in
  for _ = 0 to k do
    let prev = Array.copy prices in
    List.iter (fun (u, v, w) ->
      if prev.(u) < inf && prev.(u) + w < prices.(v) then
        prices.(v) <- prev.(u) + w
    ) flights
  done;
  if prices.(dst) = inf then -1 else prices.(dst)

File: solutions/scala/137-cheapest-flights-within-k-stops.scala
/**
 * Problem 137: Cheapest Flights Within K Stops (LeetCode 787)
 * Difficulty: Med
 * Language: Scala
 */
def findCheapestPrice(n: Int, flights: Array[Array[Int]], src: Int, dst: Int, k: Int): Int = {
  val inf = Int.MaxValue
  val init = (0 until n).map(i => if (i == src) 0 else inf).toArray
  val finalP = (0 to k).foldLeft(init) { (prices, _) =>
    flights.foldLeft(prices.clone()) { (next, f) =>
      val Array(u, v, w) = f
      if (prices(u) < inf && prices(u) + w < next(v)) { next(v) = prices(u) + w; next }
      else next
    }
  }
  if (finalP(dst) == inf) -1 else finalP(dst)
}
